renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp

```diff
@@ -1244,12 +1244,8 @@ void WrappedVulkan::vkCmdBeginRenderPass(VkCommandBuffer commandBuffer,
       if(att == NULL)
         break;
 
-      record->MarkResourceFrameReferenced(att->baseResource, eFrameRef_ReadBeforeWrite);
-      if(att->baseResourceMem != ResourceId())
-        record->MarkResourceFrameReferenced(att->baseResourceMem, eFrameRef_Read);
-      if(att->resInfo)
-        record->cmdInfo->sparse.insert(att->resInfo);
-      record->cmdInfo->dirtied.insert(att->baseResource);
+      record->MarkImageViewFrameReferenced(att, m_ImageLayouts.at(att->baseResource), ImageRange(),
```

Comment:
@baldurk, I am accessing `m_ImageLayouts` in several places to get info about the image shape. These accesses are all `const`, and do not access the mutable fields in `ImageLayouts` (tracking the current layouts and queue families) so I don't think they need to be guarded by `m_ImageLayoutsLock`. Can you verify that I'm not missing something here?

Reply:
Unfortunately maps are only thread-safe against different threads reading through `const`, but in this case a thread could be creating an image (and thus writing to it) while one of these accesses is going on, so they do need to be locked.

We could make it a read-write lock so that (rare) writes don't penalise more regular read accesses, but the original intention of the `m_ImageLayouts` was just to hold the per-subresource `VkImageLayout` (hence the name). Other metadata being tacked on there is a case of being convenient rather than the ideal place to put it.

If you need the info as widely as this I think we should reconsider where the data is stored and instead put it in the `ResourceInfo` struct accessible through the `VkResourceRecord`. Currently it's only allocated for sparse/external images for additional cold data but we could allocate it for all images. That way it's accessed via the wrapped resource itself and there's no locking or central high-frequency map.

Reply:
I see, that makes sense. 

I looked into using `ResourceInfo`, but the presence of `ResourceInfo` seems to be used as an indicator that the resource sparse (e.g. in `DescriptorSetBindingElement::AddBindRefs`). So I was a bit worried about breaking other pieces. 

Would it be correct to change the places currently checking `resInfo != NULL` (to check for sparse  resources) to instead check `resInfo != NULL && resInfo->IsSparse()`?

Reply:
Yeh that's no longer really true even in the case of external images, and would definitely need to change if we had this present for all images.

Reply:
Ok, I'm moving `layerCount`, `levelCount`, `sampleCount`, `extent`, `format` and `imageType` from `ImageLayouts` to `ResourceInfo`, and will update this PR once I get that working.